obspy/signal/cross_correlation.py

```diff
@@ -76,7 +76,44 @@ def _xcorr_slice(a, b, shift, domain='freq'):
     return c[mid - shift:mid + shift + len(c) % 2]
 
 
-def correlate(a, b, shift, demean=True, normalize=True, domain='freq'):
+def _normxcorr(a, b, shift, domain='freq'):
+    """
+    Normalized cross-correlation
+    """
+    mid = (len(a) + len(b) - 1) // 2
+    if shift is None:
+        shift = mid
+    # Normalize short window before correlation
+    b = (b - b.mean()) / (b.std() * len(b))
+    # We HAVE to pad in this case
+    a = np.pad(a, pad_width=len(b), mode='constant', constant_values=0)
+    if domain == 'time':
+        c = scipy.signal.correlate(a, b, 'valid')
+    else:
+        c = scipy.signal.fftconvolve(a, b[::-1], 'valid')
+    # Calculate moving mean and moving standard deviation
+    # using cumulative sums.
+    c = c[1:-1]
+    move_mean = np.cumsum(a)
+    move_mean = move_mean[len(b):-1] - move_mean[:-len(b) - 1]
+    numerator = c - move_mean * b.mean()
+    denominator = np.cumsum(a ** 2)
+    denominator = denominator[len(b):-1] - denominator[:-len(b) - 1]
+    np.multiply(move_mean, move_mean, out=move_mean)
+    np.divide(move_mean, len(a), out=move_mean)
+    denominator -= move_mean
+    denominator *= b.std()
+    np.maximum(denominator, 0, out=denominator)
+    np.sqrt(denominator, out=denominator)
+    # Cope with zero divisions? - scikit uses a mask
+    response = np.zeros_like(c, dtype=np.float64)
+    mask = denominator > np.finfo(np.float64).eps
```

Comment:
I don't really know a lot about this particular problem so take this comment with a grain of salt. This will be a really small number and even slightly larger numbers might results in numerically unstable calculations. Could you use something like the water level we use for instrument responses? Also is settings those values that are not in the mask to zero the correct approach?

Reply:
My take on this is that the only time you get zeros in the denominator here is when the input data are zeros - I think it is the assumption that is made in the `naive` approach too, when the `stdev` is set to 1 when it equals zero:
```python
stdev = (np.sum(a ** 2)) ** 0.5 * (np.sum(b ** 2)) ** 0.5
if stdev == 0:
    # set stdev to 1 to prevent division by 0
    # cross-correlation function will have only zeros anyway
    stdev = 1
```

It's what we do in EQcorrscan too, but we raise a warning if it happens when we don't expect it to.

By setting the limit for the mask based on the smallest resolvable 64-Bit float on the machine (the line: `mask = denominator > np.finfo(np.float64).eps`, which was a new function to me) I think we are effectively setting the appropriate water-level. I imagine there might be some cases where there is an instability, but I haven't managed to generate any so far. I also can't find a useful page to state where the limits of those divisions might be.

Reply:
Sounds reasonable to me. Worst case I guess it flips over to an `inf` and it will be immediately obvious that something is wrong - but looking at the code a bit more I guess that cannot really happen.